Mathlib/Algebra/Order/Floor/Extended.lean

```diff
@@ -0,0 +1,252 @@
+/-
+Copyright (c) 2025 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.ENat.Lattice
+import Mathlib.Data.ENNReal.Order
+import Mathlib.Data.NNReal.Basic
+import Mathlib.Data.Real.ENatENNReal
+
+/-!
+# Extended floor and ceil
+
+This file defines the extended floor and ceil functions `ENat.floor, ENat.ceil : ℝ≥0∞ → ℕ∞`.
+
+## Main declarations
+
+* `ENat.floor r`: Greatest extended natural `n` such that `n ≤ r`.
+* `ENat.ceil r`: Least extended natural `n` such that `r ≤ n`.
+
+## Notation
+
+* `⌊r⌋ₑ` is `ENat.floor r`.
+* `⌈r⌉ₑ` is `ENat.ceil r`.
+
+The index `ₑ` is used in analogy to the notation for `enorm`.
+
+## TODO
+
+The day Mathlib acquires `ENNRat`, it would be good to generalise this file to an `EFloorSemiring`
+typeclass.
+
+## Tags
+
+efloor, eceil
+-/
+
+open Set
+open scoped ENNReal NNReal
+
+namespace ENNReal
+variable {r : ℝ≥0} {n : ℕ}
+
+@[simp, norm_cast] lemma natCast_le_coe : (n : ℝ≥0∞) ≤ r ↔ n ≤ r := by simp [← coe_le_coe]
+@[simp, norm_cast] lemma coe_le_natCast : r ≤ (n : ℝ≥0∞) ↔ r ≤ n := by simp [← coe_le_coe]
+
+@[simp, norm_cast] lemma ofNNReal_add_natCast (r : ℝ≥0) (n : ℕ) : ofNNReal (r + n) = r + n := rfl
+@[simp, norm_cast] lemma ofNNReal_natCast_add (n : ℕ) (r : ℝ≥0) : ofNNReal (n + r) = n + r := rfl
+
+@[simp, norm_cast] lemma ofNNReal_sub_natCast (r : ℝ≥0) (n : ℕ) : ofNNReal (r - n) = r - n := rfl
+@[simp, norm_cast] lemma ofNNReal_natCast_sub (n : ℕ) (r : ℝ≥0) : ofNNReal (n - r) = n - r := rfl
+
+attribute [simp] sub_top
+
+end ENNReal
+
+attribute [simp] ENat.sub_top
+
+namespace ENat
+variable {r s : ℝ≥0∞} {n : ℕ∞}
+
+/-- `⌊r⌋ₑ` is the greatest extended natural `n` such that `n ≤ r`. -/
+noncomputable def floor : ℝ≥0∞ → ℕ∞
```

Comment:
Should this be defined for any `WithTop K` where `FloorSemiring K`?

Reply:
This wouldn't match `ENNReal` which isn't reducibly defeq to `WithTop NNReal`. In general, I don't think it's worth trying to fix too hard about generalisations until `ENNRat` becomes a thing

Reply:
We could make it defeq with ##23509 :)

Reply:
I am aware, but again I am not sure that it's worth thinking about the generalisation just now when there's no other application in sight yet